mitiq/factories.py

```diff
@@ -0,0 +1,130 @@
+from pyquil import Program
+from typing import List, Tuple, Callable
+
+import mitiq.qiskit.qiskit_utils as qs_utils
+import numpy as np
+
+class Factory(object):
+    """
+    This object adaptively produces new noise scaling parameters based on a historical stack of previous noise
+    scale parameters and previously calculated expectation values. Noise scaling parameters are stored on the `in_stack`
+    and the running list of expectation values are stored on the `out_stack`.
+    """
+    def __init__(self, instack: List[float], outstack: List[float]) -> None:
+        self.instack = instack
+        self.outstack = outstack
+
+    def step(self, instack: List[float], outstack: List[float]) -> None:
+        raise NotImplementedError
+
+    def is_converged(self, instack: List[float], outstack: List[float]) -> None:
+        raise NotImplementedError
+
+    def reduce(self) -> float:
+        raise NotImplementedError
+
+
+class BatchedFactory(Factory):
+    def __init__(self, scalars: List[float], instack: List[float] = None, outstack: List[float] = None) -> None:
+        """
+        Runs a series of scalar noise parameters serially.
+        :param scalars: List of scalar noise values to be executed.
+        :param instack: Running stack of noise values run so far.
+        :param outstack: Expectations calculated thus far.
+        """
+        if instack is None:
+            instack = []
+        if outstack is None:
+            outstack = []
+        super(BatchedFactory, self).__init__(instack, outstack)
+        self.scalars = scalars
+
+    def step(self, instack: List[float], outstack: List[float]) -> float:
+        try:
+            next_param = self.scalars[len(outstack)]
+        except IndexError:
+            raise IndexError(f"BatchedFactory cannot take another step. Number of batched scalars ({len(instack)}) "
+                             f"exceeded.")
+        self.instack.append(next_param)
+        return next_param
+
+    def is_converged(self, instack: List[float], outstack: List[float]) -> bool:
+        return len(instack) == len(self.scalars)
+
+# Specific extrapolation algorithms are given below.
```

Comment:
This todo list would be easier to track on the github issues rather than in the code itself. Could it be moved to the open issues(s)?

Reply:
Todo list removed